File: python/tests/string_ops.py
import time
import json
import gc

class StringOpsTest:
    def __init__(self, ops_per_iter=500000):
        self.name = "StringOps"
        self.ops_per_iter = ops_per_iter

    def workload(self):
        s = "benchmark"
        for _ in range(100):
            s += str(_)
            s = s[::-1]
            s = s.replace("a", "A")
            s.find("b")
        return len(s)

    def run_once(self, iterations):
        gc.collect()
        start = time.perf_counter()
        total_len = 0
        for _ in range(iterations // 100):
            total_len += self.workload()
        end = time.perf_counter()

        duration = end - start
        ops_done = iterations
        return {
            "time_s": duration,
            "ops": ops_done,
            "ops_per_sec": ops_done / duration if duration > 0 else 0,
            "gc_stat_after": gc.get_stats() if hasattr(gc, "get_stats") else None,
            "total_len": total_len
        }

    def run(self, runs=5, iterations=None):
        iterations = iterations or self.ops_per_iter
        results = [self.run_once(iterations) for _ in range(runs)]
        times = sorted(r["time_s"] for r in results)
        ops = sorted(r["ops_per_sec"] for r in results)

        return {
            "name": self.name,
            "runs": runs,
            "median_time_sec": times[len(times)//2],
            "median_ops_per_sec": ops[len(ops)//2],
            "raw": results
        }

if __name__ == "__main__":
    result = StringOpsTest().run()
    with open("string_ops_python.json", "w") as f:
        json.dump(result, f, indent=2)
    print(json.dumps(result, indent=2))

File: ruby/tests/string_ops.rb
require "json"

class StringOpsTest
  attr_reader :name

  def initialize(ops_per_iter = 100_000)
    @name = "StringOps"
    @ops_per_iter = ops_per_iter
  end

  def workload
    s = +"benchmark" # mutable string
    20.times do |i|
      s << i.to_s          # faster append (no reallocation)
      s.reverse!           # in-place reverse
      s.gsub!("a", "A")    # in-place substitution
      s.index("b")         # just a small operation
    end
    s.length
  end

  def run_once(iterations)
    GC.start
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    total_len = 0

    (iterations / 20).times do
      total_len += workload
    end

    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    duration = t1 - t0
    ops_done = iterations

    {
      "time_s" => duration,
      "ops" => ops_done,
      "ops_per_sec" => (duration > 0 ? ops_done / duration : 0),
      "gc_stat_after" => GC.stat,
      "total_len" => total_len
    }
  end

  def run(runs = 5, iterations = nil)
    iterations ||= @ops_per_iter
    results = Array.new(runs) { run_once(iterations) }

    times = results.map { |r| r["time_s"] }.sort
    ops = results.map { |r| r["ops_per_sec"] }.sort

    {
      "name" => @name,
      "runs" => runs,
      "median_time_sec" => times[times.length / 2],
      "median_ops_per_sec" => ops[ops.length / 2],
      "raw" => results
    }
  end
end

if __FILE__ == $0
  result = StringOpsTest.new.run
  File.write("string_ops_ruby.json", JSON.pretty_generate(result))
  puts JSON.pretty_generate(result)
end
